



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="https://dahak-metagenomics.github.io/dahak/quickstart/">
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../">
      <meta name="generator" content="mkdocs-0.17.3, mkdocs-material-2.7.2">
    
    
      
        <title>Quick Start - Dahak Metagenomics</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.8d40d89b.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.6079476c.css">
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,400i,700|Ubuntu+Mono">
        <style>body,input{font-family:"Ubuntu","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Ubuntu Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../css/custom.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="brown" data-md-color-accent="brown">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <a href="#quick-start" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://dahak-metagenomics.github.io/dahak" title="Dahak Metagenomics" class="md-header-nav__button md-logo">
          
            <i class="md-icon">line_style</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                Dahak Metagenomics
              </span>
              <span class="md-header-nav__topic">
                Quick Start
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/dahak-metagenomics/dahak" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      dahak-metagenomics/dahak
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    <span class="md-nav__button md-logo">
      
        <i class="md-icon">line_style</i>
      
    </span>
    Dahak Metagenomics
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/dahak-metagenomics/dahak" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      dahak-metagenomics/dahak
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Index" class="md-nav__link">
      Index
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../installing/" title="Installing" class="md-nav__link">
      Installing
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../running_workflows/" title="Running Workflows" class="md-nav__link">
      Running Workflows
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../config/" title="Workflow Configuration" class="md-nav__link">
      Workflow Configuration
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        Quick Start
      </label>
    
    <a href="./" title="Quick Start" class="md-nav__link md-nav__link--active">
      Quick Start
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#getting-set-up-to-run-workflows" title="Getting Set Up to Run Workflows" class="md-nav__link">
    Getting Set Up to Run Workflows
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#configuring-snakemake" title="Configuring Snakemake" class="md-nav__link">
    Configuring Snakemake
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-do-i-specify-my-data-files" title="How do I specify my data files?" class="md-nav__link">
    How do I specify my data files?
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-do-i-specify-my-workflow-configuration" title="How do I specify my workflow configuration?" class="md-nav__link">
    How do I specify my workflow configuration?
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-do-i-specify-my-workflow-parameters" title="How do I specify my workflow parameters?" class="md-nav__link">
    How do I specify my workflow parameters?
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#where-will-data-files-live" title="Where will data files live?" class="md-nav__link">
    Where will data files live?
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-really-quick-copy-and-paste-quick-start" title="The Really Quick Copy-And-Paste Quick Start" class="md-nav__link">
    The Really Quick Copy-And-Paste Quick Start
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-filtering" title="Read Filtering" class="md-nav__link">
    Read Filtering
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6">
    
    <label class="md-nav__link" for="nav-6">
      Read Filtering
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Read Filtering
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7">
    
    <label class="md-nav__link" for="nav-7">
      Assembly
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-7">
        Assembly
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-8" type="checkbox" id="nav-8">
    
    <label class="md-nav__link" for="nav-8">
      Comparison
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-8">
        Comparison
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-9" type="checkbox" id="nav-9">
    
    <label class="md-nav__link" for="nav-9">
      Taxonomic Classification
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-9">
        Taxonomic Classification
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../contrib/" title="Contributing" class="md-nav__link">
      Contributing
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#getting-set-up-to-run-workflows" title="Getting Set Up to Run Workflows" class="md-nav__link">
    Getting Set Up to Run Workflows
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#configuring-snakemake" title="Configuring Snakemake" class="md-nav__link">
    Configuring Snakemake
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-do-i-specify-my-data-files" title="How do I specify my data files?" class="md-nav__link">
    How do I specify my data files?
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-do-i-specify-my-workflow-configuration" title="How do I specify my workflow configuration?" class="md-nav__link">
    How do I specify my workflow configuration?
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-do-i-specify-my-workflow-parameters" title="How do I specify my workflow parameters?" class="md-nav__link">
    How do I specify my workflow parameters?
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#where-will-data-files-live" title="Where will data files live?" class="md-nav__link">
    Where will data files live?
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-really-quick-copy-and-paste-quick-start" title="The Really Quick Copy-And-Paste Quick Start" class="md-nav__link">
    The Really Quick Copy-And-Paste Quick Start
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#read-filtering" title="Read Filtering" class="md-nav__link">
    Read Filtering
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="quick-start">Quick Start<a class="headerlink" href="#quick-start" title="Permanent link">&para;</a></h1>
<p>Note for the <em>very</em> impatient: skip straight to <a href="#the-really-quick-copy-and-paste-quickstart">The Really Quick Copy And Paste
Quick Start</a> section.</p>
<h2 id="getting-set-up-to-run-workflows">Getting Set Up to Run Workflows<a class="headerlink" href="#getting-set-up-to-run-workflows" title="Permanent link">&para;</a></h2>
<p><strong>NOTE:</strong> This guide assumes familiarity with Dahak workflows and how they work.
For a more clear explanation of what's going on and how things work, 
start with the <a href="../running_workflows/">Running Workflows</a> page.</p>
<p>Start by cloning a copy of the repository:</p>
<div class="codehilite"><pre><span></span>$ git clone -b snakemake/comparison https://github.com/dahak-metagenomics/dahak
</pre></div>


<p>then move into the <code>workflows/</code> directory in the Dahak repository:</p>
<div class="codehilite"><pre><span></span>$ cd dahak/workflows/
</pre></div>


<p>The quick start assumes that all commands are run from the <code>workflows/</code> directory
unless stated otherwise.</p>
<p>As covered on the <a href="../running_workflows/">Running Workflows</a> page, workflows are
run by using Snakemake from the command line. The basic syntax is:</p>
<div class="codehilite"><pre><span></span>$ snakemake <span class="o">[</span>FLAGS<span class="o">]</span> &lt;target&gt;
</pre></div>


<p>(See the <a href="https://snakemake.readthedocs.io/en/stable/executable.html#all-options">executing
Snakemake</a>
page of the Snakemake documentation for a full list of options that can be used
with Snakemake.)</p>
<p>The most important flag to pass is the <code>--config</code> flag, which allows users to
specify a Snakemake configuration dictionary to customize their Dahak workflows.</p>
<p>The user should also pass the <code>--use-singularity</code> flag to tell Snakemake to use
Singularity containers. This also requires the <code>SINGULARITY_BINDPATH</code> variable
to be set, to bind-mount a local directory into the container. </p>
<p>The user should use build targets. Each workflow's build targets and details
are listed on the respective workflow's Snakemake Rules" page.</p>
<p>All together, commands to run Dahak workflows will look like this:</p>
<div class="codehilite"><pre><span></span>$ <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span> <span class="se">\</span>
        snakemake --use-singularity <span class="se">\</span>
        <span class="o">[</span>FLAGS<span class="o">]</span> &lt;target&gt;
</pre></div>


<h2 id="configuring-snakemake">Configuring Snakemake<a class="headerlink" href="#configuring-snakemake" title="Permanent link">&para;</a></h2>
<p>See <a href="../config/">Configuring Snakemake</a> for more details about Snakemake
configuration files. We cover the basics below.</p>
<h3 id="how-do-i-specify-my-data-files">How do I specify my data files?<a class="headerlink" href="#how-do-i-specify-my-data-files" title="Permanent link">&para;</a></h3>
<p>Specify the locations of your data files by assigning a key-value map
(keys are filenames, values are URLs) to the <code>files</code> key:</p>
<div class="codehilite"><pre><span></span>{
    &quot;files&quot; : {
        &quot;SRR606249_1_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0f9156c613b026430dbc7&quot;,
        &quot;SRR606249_2_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0fc7fb83f69026076be47&quot;,
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;
    }
}
</pre></div>


<p>This JSON can be put into a JSON file like <code>config/custom_readfilt.json</code> and
passed to Snakemake via the <code>--configfile</code> flag:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;data:/data&quot; \
        snakemake --configfile=config/custom_datafiles.json \
        [FLAGS] &lt;target&gt;
</pre></div>


<p>If your read files are present locally, they must be in the same
directory as the Snakemake working directory, which is specified by
the <code>data_dir</code> key (<code>data/</code> by default).</p>
<div class="admonition warning">
<p class="admonition-title">Examples in the Repository</p>
<p>Also see <a href="#"><code>workflows/config/example_datafiles.json</code></a>
in the Dahak repository, as well as the "Snakemake" page 
for each respective workflow.</p>
</div>
<h3 id="how-do-i-specify-my-workflow-configuration">How do I specify my workflow configuration?<a class="headerlink" href="#how-do-i-specify-my-workflow-configuration" title="Permanent link">&para;</a></h3>
<p>The workflow configuration is the section of the Snakemake configuration file
where you specify any information that will end up in the final file name of the
file that the workflow generates. This is generally a small number of parameters.</p>
<p>Each workflow has different build rules requiring different information.
See the "Snakemake Rules" page for a list of available build rules.
See the "Snakemake Configuration" page for a list of key-value pairs the
build rule extracts from the Snakemake configuration dictionary.</p>
<p>For example, to evaluate the quality of reads from a sequencer after quality
trimming, the <code>read_filtering_posttrim_workflow</code> Snakemake rule is used, and the
workflow configuration JSON looks like this:</p>
<div class="codehilite"><pre><span></span>{
    &quot;workflows&quot; : {
        &quot;read_filtering_posttrim_workflow&quot; : {
            &quot;sample&quot;    : [&quot;SRR606249&quot;,&quot;SRR606249_subset10&quot;],
            &quot;qual&quot;   : [&quot;2&quot;,&quot;30&quot;]
        },
    }
}
</pre></div>


<p>This file can be put into a JSON file like <code>config/custom_workflowconfig.json</code> and
passed to Snakemake via the <code>--configfile</code> flag:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;data:/data&quot; \
        snakemake --configfile=config/custom_readfilt.json \
        [FLAGS] read_filtering_posttrim_workflow
</pre></div>


<p>The <code>sample</code> list specifies the prefixes of the sample reads to 
run the read filtering workflow on. The <code>qual</code> list specifies the 
values to use for quality trimming.</p>
<div class="admonition warning">
<p class="admonition-title">Examples in the Repository</p>
<p>Also see <a href="#">workflows/config/example_workflowconfig.json</a>
in the Dahak repository, as well as the "Snakemake" page 
for each respective workflow.</p>
</div>
<h2 id="how-do-i-specify-my-workflow-parameters">How do I specify my workflow parameters?<a class="headerlink" href="#how-do-i-specify-my-workflow-parameters" title="Permanent link">&para;</a></h2>
<p>The workflow parameters section of the Snakemake configuration file
is where the details of each step of the workflow can be controlled.</p>
<p>The <strong>workflow configuration</strong> section of the Snakemake configuration dictionary
contains a subset of parameters that will end up in the final file name of the
output file generated by that workflow.  </p>
<p>The <strong>workflow parameters</strong> section of the Snakemake configuration dictionary
contains all other parameters for all intermediate steps, and is therefore
longer and contains more options than the workflow configuration section.</p>
<p>The workflow parameters are organized by workflow, with one additional
parameters section for setting the version and URL of the biocontainers 
used in each workflow.</p>
<p>For example, the read filtering workflow parameters section 
is under the top-level <code>read_filtering</code> key:</p>
<div class="codehilite"><pre><span></span>{
    &quot;read_filtering&quot; : {
        &quot;read_patterns&quot; : {
            &quot;pre_trimming_pattern&quot;  : &quot;{sample}_{direction}_reads.fq.gz&quot;,
            &quot;post_trimming_pattern&quot; : &quot;{sample}_{direction}.trim{qual}.fq.gz&quot;,
        },
        &quot;direction_labels&quot; : {
            &quot;forward&quot; : &quot;1&quot;,
            &quot;reverse&quot; : &quot;2&quot;
        },
        &quot;quality_assessment&quot; : {
            &quot;fastqc_suffix&quot;: &quot;fastqc&quot;,
        },
        &quot;quality_trimming&quot; : {
            &quot;trim_suffix&quot; : &quot;se&quot;
        },
        &quot;interleaving&quot; : {
            &quot;interleave_suffix&quot; : &quot;pe&quot;
        },
        &quot;adapter_file&quot; : {
            &quot;name&quot; : &quot;TruSeq2-PE.fa&quot;,
            &quot;url&quot;  : &quot;http://dib-training.ucdavis.edu.s3.amazonaws.com/mRNAseq-semi-2015-03-04/TruSeq2-PE.fa&quot;
        }
    }
}
</pre></div>


<p>This file can be put into a JSON file like <code>config/custom_readfilt.json</code> and
passed to Snakemake via the <code>--configfile</code> flag:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;data:/data&quot; \
        snakemake --configfile=config/custom_readfilt.json \
        [FLAGS] read_filtering_posttrim_workflow
</pre></div>


<p>(See the <a href="../readfilt_snakemake/">Read Filtering Snakemake</a> page
for an explanation of each of the above options.)</p>
<div class="admonition warning">
<p class="admonition-title">Examples in the Repository</p>
<p>The <a href="#"><code>workflows/config/example_workflowparams.json</code></a>
file in the repository contains an example, but JSON files
do not contain any comments. For a well-commented version,
check the `workflows/config/default_workflowparams.settings](#)
file.</p>
</div>
<h2 id="where-will-data-files-live">Where will data files live?<a class="headerlink" href="#where-will-data-files-live" title="Permanent link">&para;</a></h2>
<p>To set the scratch/working directory for the Snakemake workflows, 
in which all intermediate and final data files will be placed, 
set the <code>data_dir</code> key in the Snakemake configuration dictionary. 
If this option is not set, it will be <code>data</code> by default.  No 
trailing slash is needed when setting this option.</p>
<div class="admonition warning">
<p class="admonition-title">Important Note About <code>data_dir</code></p>
<p>If you use a custom directory, you must also 
adjust the <code>SINGULARITY_BINDPATH</code> variable accordingly.</p>
</div>
<p>For example, to put all intermediate files into the <code>work/</code> directory
instead of the <code>data/</code> directory, you can use the following JSON
file:</p>
<div class="codehilite"><pre><span></span>{
    &quot;data_dir&quot; : &quot;work&quot;
}
</pre></div>


<p>This file can be put into a JSON file like <code>config/custom_datadir.json</code> and
passed to Snakemake via the <code>--configfile</code> flag:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;work:/work&quot; \
        snakemake --configfile=config/custom_datadir.settings \
        [FLAGS] &lt;target&gt;
</pre></div>


<h2 id="the-really-quick-copy-and-paste-quick-start">The Really Quick Copy-And-Paste Quick Start<a class="headerlink" href="#the-really-quick-copy-and-paste-quick-start" title="Permanent link">&para;</a></h2>
<p>Now that we've provided some examples that you can use, let's run through
the entire process start to finish to illustrate how this works.</p>
<h3 id="read-filtering">Read Filtering<a class="headerlink" href="#read-filtering" title="Permanent link">&para;</a></h3>
<p>We will run two variations of the read filtering workflow, and perform a quality
assessment of our reads both before and after quality trimming.</p>
<p>Before you begin, make sure you have everything listed on the
<a href="../installing/">Installing</a> page available on your command line.</p>
<p>Start by cloning a copy of the repository:</p>
<div class="codehilite"><pre><span></span>$ git clone -b snakemake/comparison https://github.com/dahak-metagenomics/dahak
</pre></div>


<p>then move into the <code>workflows/</code> directory of the Dahak repository:</p>
<div class="codehilite"><pre><span></span>$ cd dahak/workflows/
</pre></div>


<p>Now create a configuration JSON file that:</p>
<ul>
<li>Provides URLs at which each read filtering file can be accessed</li>
<li>Provides a set of quality trimming values to use (2 and 30)</li>
<li>Sets all read filtering parameters (for simplicity, we will set each
    parameter to its default value)</li>
</ul>
<div class="codehilite"><pre><span></span>{
    &quot;files&quot; : {
        &quot;SRR606249_1_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0f9156c613b026430dbc7&quot;,
        &quot;SRR606249_2_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0fc7fb83f69026076be47&quot;,
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
        &quot;SRR606249_subset25_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1039a594d900263120c38&quot;,
        &quot;SRR606249_subset25_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f104ed594d90026411f486&quot;
    },

    &quot;workflows&quot; : {
        &quot;read_filtering_pretrim_workflow&quot; : {
            &quot;sample&quot;    : [&quot;SRR606249_subset25&quot;,&quot;SRR606249_subset10&quot;],
        },
        &quot;read_filtering_posttrim_workflow&quot; : {
            &quot;sample&quot;    : [&quot;SRR606249_subset25&quot;,&quot;SRR606249_subset10&quot;],
            &quot;qual&quot;   : [&quot;2&quot;,&quot;30&quot;]
        },
    },

    &quot;read_filtering&quot; : {
        &quot;read_patterns&quot; : {
            &quot;pre_trimming_pattern&quot;  : &quot;{sample}_{direction}_reads.fq.gz&quot;,
            &quot;post_trimming_pattern&quot; : &quot;{sample}_{direction}.trim{qual}.fq.gz&quot;,
        },
        &quot;direction_labels&quot; : {
            &quot;forward&quot; : &quot;1&quot;,
            &quot;reverse&quot; : &quot;2&quot;
        },
        &quot;quality_assessment&quot; : {
            &quot;fastqc_suffix&quot;: &quot;fastqc&quot;,
        },
        &quot;quality_trimming&quot; : {
            &quot;trim_suffix&quot; : &quot;se&quot;
        },
        &quot;interleaving&quot; : {
            &quot;interleave_suffix&quot; : &quot;pe&quot;
        },
        &quot;adapter_file&quot; : {
            &quot;name&quot; : &quot;TruSeq2-PE.fa&quot;,
            &quot;url&quot;  : &quot;http://dib-training.ucdavis.edu.s3.amazonaws.com/mRNAseq-semi-2015-03-04/TruSeq2-PE.fa&quot;
        }
    }
}
</pre></div>


<p>Put this file into <code>config/custom_readfilt_workflow.json</code> (in the <code>workflows</code>
directory of the repository). We want to run two workflows: one pre-trimming
quality assessment, and one post-trimming quality assessment.</p>
<p>We add two flags to the Snakemake command: <code>-n</code> and <code>-p</code>.</p>
<p><code>-n</code> does a <em>dry-run</em>, meaning Snakemake will print out what tasks it is
going to run, but will not acutally run them.</p>
<p><code>-p</code> tells Snakemake to print the shell commands that it will run for each
workflow step. This is useful for understanding what Snakemake is doing
and what commands and options are being run.</p>
<div class="codehilite"><pre><span></span>$ export SINGULARITY_BINDPATH=&quot;data:/data&quot;

$ snakemake -p -n \
        --configfile=config/custom_readfilt_workflow.json \
        read_filtering_pretrim_workflow

$ snakemake -p -n \
        --configfile=config/custom_readfilt_workflow.json \
        read_filtering_posttrim_workflow
</pre></div>


<p>Once we have reviewed the output from Snakemake and are satisfied
it is running the correct workflow and commands, we can actually
run the workflow by removing the <code>-n</code> flag: </p>
<div class="codehilite"><pre><span></span>$ export SINGULARITY_BINDPATH=&quot;data:/data&quot; 

$ snakemake -p \
        --configfile=config/custom_readfilt_workflow.json \
        read_filtering_pretrim_workflow

$ snakemake -p \
        --configfile=config/custom_readfilt_workflow.json \
        read_filtering_posttrim_workflow
</pre></div>


<p>We can also run both workflows at once by specifying two targets:</p>
<div class="codehilite"><pre><span></span>$ snakemake -p \
        --configfile=config/custom_readfilt_workflow.json \
        read_filtering_pretrim_workflow read_filtering_posttrim_workflow
</pre></div>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../config/" title="Workflow Configuration" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Workflow Configuration
              </span>
            </div>
          </a>
        
        
          <a href="../readfilt_overview/" title="Overview" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Overview
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2017, The Regents of the University of California.<br /> Written and maintained by the <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a> at UC Davis.<br /> Released under the <a href="https://opensource.org/licenses/BSD-3-Clause">BSD 3-Clause License</a>.<br /><br />
          </div>
        
        powered by
        <a href="http://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.0cf9b500.js"></script>
      
      <script>app.initialize({version:"0.17.3",url:{base:".."}})</script>
      
        <script src="../search/require.js"></script>
      
        <script src="../search/search.js"></script>
      
    
    
      
    
  </body>
</html>