



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="https://dahak-metagenomics.github.io/dahak/quickstart/">
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../">
      <meta name="generator" content="mkdocs-0.17.3, mkdocs-material-2.7.2">
    
    
      
        <title>Quick Start - Dahak Metagenomics</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.8d40d89b.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.6079476c.css">
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,400i,700|Ubuntu+Mono">
        <style>body,input{font-family:"Ubuntu","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Ubuntu Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../css/custom.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="brown" data-md-color-accent="brown">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <a href="#quick-start" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://dahak-metagenomics.github.io/dahak" title="Dahak Metagenomics" class="md-header-nav__button md-logo">
          
            <i class="md-icon">line_style</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                Dahak Metagenomics
              </span>
              <span class="md-header-nav__topic">
                Quick Start
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/dahak-metagenomics/dahak" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      dahak-metagenomics/dahak
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    <span class="md-nav__button md-logo">
      
        <i class="md-icon">line_style</i>
      
    </span>
    Dahak Metagenomics
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/dahak-metagenomics/dahak" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      dahak-metagenomics/dahak
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Index" class="md-nav__link">
      Index
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../installing/" title="Installing" class="md-nav__link">
      Installing
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../running_workflows/" title="Running Workflows" class="md-nav__link">
      Running Workflows
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../config/" title="Workflow Configuration" class="md-nav__link">
      Workflow Configuration
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
    <a href="./" title="Quick Start" class="md-nav__link md-nav__link--active">
      Quick Start
    </a>
    
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6">
    
    <label class="md-nav__link" for="nav-6">
      Read Filtering
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Read Filtering
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_snakemake/" title="Snakemake Rules" class="md-nav__link">
      Snakemake Rules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_config/" title="Snakemake Configuration" class="md-nav__link">
      Snakemake Configuration
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7">
    
    <label class="md-nav__link" for="nav-7">
      Assembly
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-7">
        Assembly
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_snakemake/" title="Snakemake Rules" class="md-nav__link">
      Snakemake Rules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_config/" title="Snakemake Configuration" class="md-nav__link">
      Snakemake Configuration
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-8" type="checkbox" id="nav-8">
    
    <label class="md-nav__link" for="nav-8">
      Comparison
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-8">
        Comparison
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_snakemake/" title="Snakemake Rules" class="md-nav__link">
      Snakemake Rules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_config/" title="Snakemake Configuration" class="md-nav__link">
      Snakemake Configuration
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-9" type="checkbox" id="nav-9">
    
    <label class="md-nav__link" for="nav-9">
      Taxonomic Classification
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-9">
        Taxonomic Classification
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_snakemake/" title="Snakemake Rules" class="md-nav__link">
      Snakemake Rules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_config/" title="Snakemake Configuration" class="md-nav__link">
      Snakemake Configuration
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../contrib/" title="Contributing" class="md-nav__link">
      Contributing
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="quick-start">Quick Start<a class="headerlink" href="#quick-start" title="Permanent link">&para;</a></h1>
<h1 id="getting-set-up-to-run-workflows">Getting Set Up to Run Workflows<a class="headerlink" href="#getting-set-up-to-run-workflows" title="Permanent link">&para;</a></h1>
<h2 id="how-do-i-run-workflows">How do I run workflows?<a class="headerlink" href="#how-do-i-run-workflows" title="Permanent link">&para;</a></h2>
<p>Start by cloning a copy of the repository:</p>
<div class="codehilite"><pre><span></span>$ git clone https://github.com/dahak-metagenomics/dahak
</pre></div>


<p>then move into the <code>workflows/</code> directory in the Dahak repository:</p>
<div class="codehilite"><pre><span></span>$ cd dahak/workflows/
</pre></div>


<p>As covered on the <a href="../running_workflows/">Running Workflows</a> page, workflows are
run by using Snakemake from the command line. The basic syntax is:</p>
<div class="codehilite"><pre><span></span>$ snakemake <span class="o">[</span>FLAGS<span class="o">]</span> &lt;target&gt;
</pre></div>


<p>(See the <a href="https://snakemake.readthedocs.io/en/stable/executable.html#all-options">executing
Snakemake</a>
page of the Snakemake documentation for a full list of options that can be used
with Snakemake.)</p>
<p>Passing the <code>--config</code> flag to Snakemake allows users to specify configuration
variables that are passed on to Snakemake and used to assemble and carry out
workflow tasks.</p>
<p>To run via Singularity, set the <code>SINGULARLITY_BINDPATH</code> variable (below we tell
Singularity to bind-mount the local <code>data/</code> directory into the container at <code>/data</code>)
and pass the <code>--use-singularity</code> flag to Snakemake:</p>
<div class="codehilite"><pre><span></span>$ <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span> snakemake --use-singularity <span class="se">\</span>
        <span class="o">[</span>FLAGS<span class="o">]</span> &lt;target&gt;
</pre></div>


<h2 id="how-do-i-customize-my-workflow">How do I customize my workflow?<a class="headerlink" href="#how-do-i-customize-my-workflow" title="Permanent link">&para;</a></h2>
<p>You can customize your workflow using the Snakemake configuration dictionary.
This can be set by setting key-value pairs in a JSON file and passing the JSON
file to Snakemake via the <code>--config</code> command line option:</p>
<div class="codehilite"><pre><span></span>$ snakemake --config=config/custom_datafiles.json  [FLAGS]  &lt;targets&gt;
</pre></div>


<p>See the <a href="../config/">Workflow Configuration</a> page or the examples below 
for details.</p>
<h2 id="how-do-i-specify-my-data-files">How do I specify my data files?<a class="headerlink" href="#how-do-i-specify-my-data-files" title="Permanent link">&para;</a></h2>
<p>To set the names and URLs of read files, set the <code>files</code> key in the Snakemake
configuration dictionary to a list of key-value pairs, where the key is the 
name of the read file and the value is the URL of the file (do not include 
<code>http://</code> or <code>https://</code> in the URL).</p>
<p>For example, the following JSON block will provide a list of reads and their
corresponding URLs:</p>
<div class="codehilite"><pre><span></span>{
    &quot;files&quot; : {
        &quot;SRR606249_1_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0f9156c613b026430dbc7&quot;,
        &quot;SRR606249_2_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0fc7fb83f69026076be47&quot;,
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
        &quot;SRR606249_subset25_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1039a594d900263120c38&quot;,
        &quot;SRR606249_subset25_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f104ed594d90026411f486&quot;,
        &quot;SRR606249_subset50_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1082d6c613b026430e5cf&quot;,
        &quot;SRR606249_subset50_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10ac6594d900262123e77&quot;,
    }
}
</pre></div>


<p>This can be placed in a JSON file like <code>dahak/workflows/config/custom_datafiles.json</code> 
and passed to Snakemake using the <code>--config</code> flag like:</p>
<div class="codehilite"><pre><span></span>$ snakemake --config=config/custom_datafiles.json \
        [FLAGS] &lt;target&gt;
</pre></div>


<p><strong>NOTE:</strong> The read trimming workflow assumes the read files are not available
locally and uses a rule to download them from the given URL using <code>wget</code> if they
are not present.  If your read files <em>are</em> present locally, they must be placed
in the working directory (<code>data/</code> by default, also see the "Where will data
files live?" section below). If the read files are present, Snakemake will not
run the rule to download read files, so you can use an empty string for the URL.</p>
<h2 id="what-targets-do-i-use">What targets do I use?<a class="headerlink" href="#what-targets-do-i-use" title="Permanent link">&para;</a></h2>
<p>Each workflow has a set of "build rules" that will trigger
all rules required to run a particular workflow.</p>
<p>Each workflow lists the available build rules on the respective
"Snakemake Rules" page for that workflow (see left side navigation menu).
The build rules require some information about which read files
to run the workflow on; the information required is covered on
each "Snakemake Rules" page.</p>
<p>We also cover usage of several build rules below.</p>
<h2 id="how-do-i-specify-workflow-parameters">How do I specify workflow parameters?<a class="headerlink" href="#how-do-i-specify-workflow-parameters" title="Permanent link">&para;</a></h2>
<p>The default workflow parameter values are set in <code>default_workflowparams.settings</code>.
Any of these values can be overridden using a custom JSON file, as described
above and on the <a href="../config/">Workflow Configuration</a> page. For example,
the default version of trimmomatic used is 0.36, obtained from the biocontainers
project. To override the version of trimmomatic, the following JSON file
could be used:</p>
<div class="codehilite"><pre><span></span>{
    &quot;biocontainers&quot; : {
        &quot;trimmomatic&quot; : {
            &quot;use_local&quot; : false,
            &quot;quayurl&quot; : &quot;quay.io/biocontainers/trimmomatic&quot;,
            &quot;version&quot; : &quot;0.38--5&quot;
        }
    }
}
</pre></div>


<p>This can be placed in a JSON file like <code>dahak/workflows/config/custom_workflowparams.json</code> 
and passed to Snakemake using the <code>--config</code> flag like:</p>
<div class="codehilite"><pre><span></span>$ snakemake --config=config/custom_workflowparams.json \
        [FLAGS] &lt;target&gt;
</pre></div>


<h2 id="where-will-data-files-live">Where will data files live?<a class="headerlink" href="#where-will-data-files-live" title="Permanent link">&para;</a></h2>
<p>To set the scratch/working directory for the Snakemake workflows, 
in which all intermediate and final data files will be placed, 
set the <code>data_dir</code> key in the Snakemake configuration dictionary. 
If this option is not set, it will be <code>data</code> by default.</p>
<p><strong>IMPORTANT:</strong> If you use a custom directory, you must also 
adjust the <code>SINGULARITY_BINDPATH</code> variable accordingly.</p>
<p><strong>NOTE:</strong> No trailing slash is needed.</p>
<p>For example, to put all intermediate files into the <code>work/</code> directory
instead of the <code>data/</code> directory, you can use the following JSON
file:</p>
<div class="codehilite"><pre><span></span>{
    &quot;data_dir&quot; : &quot;work&quot;
}
</pre></div>


<p>This can be used by passing the <code>--configfile</code> flag to Snakemake
and updating the Singularity environment variable:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;work:/work&quot; \
        snakemake --configfile=config/custom_scratch.settings \
        [FLAGS] &lt;target&gt;
</pre></div>


<h2 id="how-do-i-specify-which-samples-to-run-the-workflow-on">How do I specify which samples to run the workflow on?<a class="headerlink" href="#how-do-i-specify-which-samples-to-run-the-workflow-on" title="Permanent link">&para;</a></h2>
<p>Each workflow has a set of build rules that will trigger all other rules
required to run a workflow. These build rules require the user to specify
which read files to begin with. We cover several examples below.</p>
<p><br />
<br /></p>
<h2 id="quick-start-read-filtering">Quick Start: Read Filtering<a class="headerlink" href="#quick-start-read-filtering" title="Permanent link">&para;</a></h2>
<p>To run the read filtering workflow, there are several workflow configuration
options and workflow parameter values that the user should set.</p>
<p>The user must start by specifying the location of their read files as covered in
the "How do I specify my data files?" section above.</p>
<p>Workflow configuration options the user must specify:</p>
<ul>
<li>List of samples to run the read filtering workflow on</li>
<li>List of quality values to use when performing read filtering</li>
</ul>
<p>Workflow parameters the user should specify:</p>
<ul>
<li>Pre-trimming and post-trimming filename patterns</li>
</ul>
<p>The read filtering workflow requires that the user specify
the following:</p>
<ul>
<li>List of read files and corresponding URLs;</li>
<li>List of quality values to use when performing read filtering;</li>
<li>Pre-trimming and post-trimming filename patterns;</li>
<li>Adapter file name and URL</li>
</ul>
<p>These can be specified in a configuration file in JSON format
as follows:</p>
<p><strong><code>config/read_filtering_config.settings</code>:</strong></p>
<div class="codehilite"><pre><span></span>{
    #
    # Specify where to get the read files
    # 
    &quot;files&quot; : {
        #
        # URLs for full and subsampled reads
        # these must match the pre_trimming pattern given below
        &quot;SRR606249_1_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0f9156c613b026430dbc7&quot;,
        &quot;SRR606249_2_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0fc7fb83f69026076be47&quot;,
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
        &quot;SRR606249_subset25_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1039a594d900263120c38&quot;,
        &quot;SRR606249_subset25_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f104ed594d90026411f486&quot;,
        &quot;SRR606249_subset50_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1082d6c613b026430e5cf&quot;,
        &quot;SRR606249_subset50_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10ac6594d900262123e77&quot;,
    },


    #
    # Specify how to do read filtering
    # 
    &quot;read_filtering&quot; : {
        # 
        # The workflow actually builds the rules to download 
        # the read_files by using the pre_trimming_pattern.
        &quot;read_patterns&quot; : {
            #
            # filename pattern for pre-trimmed reads
            # Note: the files section (top) listing URLS for read files
            # MUST match the pre_trimmming_pattern.
            &quot;pre_trimming_pattern&quot;  : &quot;{sample}_{direction}_reads.fq.gz&quot;,
            #
            # filename pattern for post-trimmed reads
            &quot;post_trimming_pattern&quot; : &quot;{sample}_{direction}.trim{qual}.fq.gz&quot;,
        },
        #
        # Set the read adapter file
        &quot;adapter_file&quot; : {
            # 
            # name and URL for the sequencer adapter file
            &quot;name&quot; : &quot;TruSeq2-PE.fa&quot;,
            &quot;url&quot;  : &quot;http://dib-training.ucdavis.edu.s3.amazonaws.com/mRNAseq-semi-2015-03-04/TruSeq2-PE.fa&quot;
        }
    },


    #
    # Specify what reads to filter
    # 
    &quot;workflows&quot; : {

        &quot;read_filtering_workflow&quot; : {
            #
            # These parameters determine which samples
            # the read filtering workflow will be run on.
            &quot;sample&quot;    : [&quot;SRR606249_subset10&quot;,&quot;SRR606249_subset25&quot;],
            &quot;direction&quot; : [&quot;1&quot;,&quot;2&quot;],
            &quot;quality&quot;   : [&quot;2&quot;,&quot;30&quot;],
        }
    }
}
</pre></div>


<p>To run the read filtering workflow, call snakemake and pass the
<code>read_filtering_workflow</code> build rule:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;data:/data&quot; \
    snakemake \
    --use-singularity \
    --configfile=config/read_filtering_config.settings \
    read_filtering_workflow
</pre></div>


<p>When this command is run, it uses the filename pattern for the 
post-trimming step to construct filenames resulting from the
read filtering step, and asks Snakemake to run the rules that
produce those files.</p>
<p>More information about the read filtering workflow:</p>
<ul>
<li><a href="../readfilt_walkthru/">Read Filtering Walkthrough</a></li>
<li><a href="../readfilt_snakemake/">Read Filtering Snakemake Rules</a></li>
<li><a href="../readfilt_config/">Read Filtering Parameters and Configuration</a></li>
</ul>
<p><br />
<br /></p>
<h2 id="quick-start-taxonomic-classification">Quick Start: Taxonomic Classification<a class="headerlink" href="#quick-start-taxonomic-classification" title="Permanent link">&para;</a></h2>
<p>The taxonomic classification workflow builds on the read filtering
workflow, so the user should specify all of the information
listed in the "Quick Start: Read Filtering" section (above). 
In addition, the user should specify the following:</p>
<ul>
<li>Percent threshold to use for taxa filtering</li>
<li>Taxonomic rank to use for kaiju2krona</li>
<li>Kaiju database name and URL</li>
</ul>
<p>These can be specified in a configuration file in JSON format
as follows:</p>
<p><strong><code>config/taxonomic_classification_config.settings</code>:</strong></p>
<div class="codehilite"><pre><span></span>{
    #
    # Specify where to get the read files
    # 
    &quot;files&quot; : {
        #
        # URLs for full and subsampled reads
        # these must match the pre_trimming pattern given below
        &quot;SRR606249_1_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0f9156c613b026430dbc7&quot;,
        &quot;SRR606249_2_reads.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0fc7fb83f69026076be47&quot;,
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
        &quot;SRR606249_subset25_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1039a594d900263120c38&quot;,
        &quot;SRR606249_subset25_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f104ed594d90026411f486&quot;,
        &quot;SRR606249_subset50_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1082d6c613b026430e5cf&quot;,
        &quot;SRR606249_subset50_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10ac6594d900262123e77&quot;,
    },


    #
    # Specify how to do taxonomic classification
    # 
    &quot;taxonomic_classification&quot; : {

        &quot;filter_taxa&quot; : {
            #
            # percent threshold for taxa filtering
            &quot;pct_threshold&quot; : 1
        },

        &quot;kaiju2krona&quot; : {
            #
            # specify the taxonomic rank for kaiju2krona to use
            &quot;taxonomic_rank&quot; : &quot;genus&quot;
        },

        &quot;kaiju&quot; : {
            &quot;dmp1&quot; : &quot;nodes.dmp&quot;,
            &quot;dmp2&quot; : &quot;names.dmp&quot;,
            &quot;fmi&quot;  : &quot;kaiju_db_nr_euk.fmi&quot;,
            &quot;tar&quot;  : &quot;kaiju_index_nr_euk.tgz&quot;,
            #&quot;url&quot;  : &quot;https://s3.amazonaws.com/dahak-project-ucdavis/kaiju&quot;,
            &quot;url&quot;  : &quot;http://kaiju.binf.ku.dk/database&quot;,
            &quot;out&quot;  : &quot;{sample}.kaiju_output.trim{qual}.out&quot;
        },

        &quot;reads&quot; : {
            # set filename pattern for reads
            &quot;fq_fwd&quot; : &quot;{sample}_1.trim{qual}.fq.gz&quot;,
            &quot;fq_rev&quot; : &quot;{sample}_2.trim{qual}.fq.gz&quot;
        },

        &quot;sourmash&quot; : { 
            #
            # URL base for SBT tree
            &quot;sbturl&quot;  : &quot;s3-us-west-1.amazonaws.com/spacegraphcats.ucdavis.edu&quot;,
            # 
            # name of SBT tar file
            &quot;sbttar&quot;  : &quot;microbe-{database}-sbt-k{ksize}-2017.05.09.tar.gz&quot;,
            #
            # name of SBT file when unpacked
            &quot;sbtunpack&quot; : &quot;{database}-k{ksize}.sbt.json&quot;
        },

        &quot;visualize_krona&quot; : {
            #
            # .summary will be replaced with .html for the final report
            &quot;input_summary&quot;  : &quot;{sample}.kaiju_output.trim{qual}.summary&quot;,
        }
    },


    ###########################################
    # Everything below this line is the same
    # as the read filtering example above.
    ###########################################


    #
    # Specify how to do read filtering
    # 
    &quot;read_filtering&quot; : {
        # 
        # The workflow actually builds the rules to download 
        # the read_files by using the pre_trimming_pattern.
        &quot;read_patterns&quot; : {
            #
            # filename pattern for pre-trimmed reads
            # Note: the files section (top) listing URLS for read files
            # MUST match the pre_trimmming_pattern.
            &quot;pre_trimming_pattern&quot;  : &quot;{sample}_{direction}_reads.fq.gz&quot;,
            #
            # filename pattern for post-trimmed reads
            &quot;post_trimming_pattern&quot; : &quot;{sample}_{direction}.trim{qual}.fq.gz&quot;,
        },
        #
        # Set the read adapter file
        &quot;adapter_file&quot; : {
            # 
            # name and URL for the sequencer adapter file
            &quot;name&quot; : &quot;TruSeq2-PE.fa&quot;,
            &quot;url&quot;  : &quot;http://dib-training.ucdavis.edu.s3.amazonaws.com/mRNAseq-semi-2015-03-04/TruSeq2-PE.fa&quot;
        }
    },


    #
    # Specify what reads to filter
    # 
    &quot;workflows&quot; : {

        &quot;read_filtering_workflow&quot; : {
            #
            # These parameters determine which samples
            # the read filtering workflow will be run on.
            &quot;sample&quot;    : [&quot;SRR606249_subset10&quot;,&quot;SRR606249_subset25&quot;],
            &quot;direction&quot; : [&quot;1&quot;,&quot;2&quot;],
            &quot;quality&quot;   : [&quot;2&quot;,&quot;30&quot;],
        }
    }
}
</pre></div>


<p>To run the read filtering workflow, call snakemake and pass the
<code>taxonomic_classification_workflow</code> build rule:</p>
<div class="codehilite"><pre><span></span>$ SINGULARITY_BINDPATH=&quot;data:/data&quot; \
    snakemake \
    --use-singularity \
    --configfile=config/taxonomic_classification_config.settings \
    taxonomic_classification_workflow
</pre></div>


<p>More information about the taxonomic classification workflow:</p>
<ul>
<li><a href="../taxclass_walkthru/">Taxonomic Classification Walkthrough</a></li>
<li><a href="../taxclass_snakemake/">Taxonomic Classification Snakemake Rules</a></li>
<li><a href="../taxclass_config/">Taxonomic Classification Parameters and Configuration</a></li>
</ul>
<p><br />
<br /></p>
<h2 id="quick-start-assembly">Quick Start: Assembly<a class="headerlink" href="#quick-start-assembly" title="Permanent link">&para;</a></h2>
<p>More information about the assembly workflow:</p>
<ul>
<li><a href="../assembly_walkthru/">Assembly Walkthrough</a></li>
<li><a href="../assembly_snakemake/">Assembly Snakemake Rules</a></li>
<li><a href="../assembly_config/">Assembly Parameters and Configuration</a></li>
</ul>
<p><br />
<br /></p>
<h2 id="quick-start-comparison">Quick Start: Comparison<a class="headerlink" href="#quick-start-comparison" title="Permanent link">&para;</a></h2>
<p>More information about the workflow:</p>
<ul>
<li><a href="../comparison_walkthru/">Comparison Walkthrough</a></li>
<li><a href="../comparison_snakemake/">Comparison Snakemake Rules</a></li>
<li><a href="../comparison_config/">Comparison Parameters and Configuration</a></li>
</ul>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../config/" title="Workflow Configuration" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Workflow Configuration
              </span>
            </div>
          </a>
        
        
          <a href="../readfilt_overview/" title="Overview" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Overview
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2017, The Regents of the University of California.<br /> Written and maintained by the <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a> at UC Davis.<br /> Released under the <a href="https://opensource.org/licenses/BSD-3-Clause">BSD 3-Clause License</a>.<br /><br />
          </div>
        
        powered by
        <a href="http://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.0cf9b500.js"></script>
      
      <script>app.initialize({version:"0.17.3",url:{base:".."}})</script>
      
        <script src="../search/require.js"></script>
      
        <script src="../search/search.js"></script>
      
    
    
      
    
  </body>
</html>