



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="https://dahak-metagenomics.github.io/dahak/quickstart/">
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../">
      <meta name="generator" content="mkdocs-0.17.5, mkdocs-material-2.7.2">
    
    
      
        <title>Quick Start - Dahak Metagenomics</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.8d40d89b.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.6079476c.css">
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,400i,700|Ubuntu+Mono">
        <style>body,input{font-family:"Ubuntu","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Ubuntu Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../css/custom.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="brown" data-md-color-accent="brown">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <a href="#quick-start" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://dahak-metagenomics.github.io/dahak" title="Dahak Metagenomics" class="md-header-nav__button md-logo">
          
            <i class="md-icon">line_style</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                Dahak Metagenomics
              </span>
              <span class="md-header-nav__topic">
                Quick Start
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/dahak-metagenomics/dahak" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      dahak-metagenomics/dahak
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    <span class="md-nav__button md-logo">
      
        <i class="md-icon">line_style</i>
      
    </span>
    Dahak Metagenomics
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/dahak-metagenomics/dahak" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      dahak-metagenomics/dahak
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Index" class="md-nav__link">
      Index
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../installing/" title="Installing" class="md-nav__link">
      Installing
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../running_workflows/" title="Running Workflows" class="md-nav__link">
      Running Workflows
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../config/" title="Workflow Configuration" class="md-nav__link">
      Workflow Configuration
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        Quick Start
      </label>
    
    <a href="./" title="Quick Start" class="md-nav__link md-nav__link--active">
      Quick Start
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#useful-snakemake-flags" title="Useful Snakemake Flags" class="md-nav__link">
    Useful Snakemake Flags
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-filtering" title="Read Filtering" class="md-nav__link">
    Read Filtering
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly" title="Assembly" class="md-nav__link">
    Assembly
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#comparison" title="Comparison" class="md-nav__link">
    Comparison
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomic-classification" title="Taxonomic Classification" class="md-nav__link">
    Taxonomic Classification
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#taxonomic-classification-with-sourmash" title="Taxonomic Classification with Sourmash" class="md-nav__link">
    Taxonomic Classification with Sourmash
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#signatures-workflow" title="Signatures Workflow" class="md-nav__link">
    Signatures Workflow
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gather-workflow" title="Gather Workflow" class="md-nav__link">
    Gather Workflow
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#taxonomic-classification-with-kaiju" title="Taxonomic Classification with Kaiju" class="md-nav__link">
    Taxonomic Classification with Kaiju
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kaiju-report-workflow" title="Kaiju Report Workflow" class="md-nav__link">
    Kaiju Report Workflow
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaiju-filtered-species-report-workflow" title="Kaiju Filtered Species Report Workflow" class="md-nav__link">
    Kaiju Filtered Species Report Workflow
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaiju-filtered-species-by-class-report-workflow" title="Kaiju Filtered Species by Class Report Workflow" class="md-nav__link">
    Kaiju Filtered Species by Class Report Workflow
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../troubleshooting/" title="Troubleshooting" class="md-nav__link">
      Troubleshooting
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7">
    
    <label class="md-nav__link" for="nav-7">
      Read Filtering
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-7">
        Read Filtering
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../readfilt_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-8" type="checkbox" id="nav-8">
    
    <label class="md-nav__link" for="nav-8">
      Assembly
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-8">
        Assembly
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assembly_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-9" type="checkbox" id="nav-9">
    
    <label class="md-nav__link" for="nav-9">
      Comparison
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-9">
        Comparison
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../comparison_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-10" type="checkbox" id="nav-10">
    
    <label class="md-nav__link" for="nav-10">
      Taxonomic Classification
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-10">
        Taxonomic Classification
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_overview/" title="Overview" class="md-nav__link">
      Overview
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_walkthru/" title="Walkthrough" class="md-nav__link">
      Walkthrough
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../taxclass_snakemake/" title="Snakemake" class="md-nav__link">
      Snakemake
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../contrib/" title="Contributing" class="md-nav__link">
      Contributing
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#useful-snakemake-flags" title="Useful Snakemake Flags" class="md-nav__link">
    Useful Snakemake Flags
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-filtering" title="Read Filtering" class="md-nav__link">
    Read Filtering
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly" title="Assembly" class="md-nav__link">
    Assembly
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#comparison" title="Comparison" class="md-nav__link">
    Comparison
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomic-classification" title="Taxonomic Classification" class="md-nav__link">
    Taxonomic Classification
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#taxonomic-classification-with-sourmash" title="Taxonomic Classification with Sourmash" class="md-nav__link">
    Taxonomic Classification with Sourmash
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#signatures-workflow" title="Signatures Workflow" class="md-nav__link">
    Signatures Workflow
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gather-workflow" title="Gather Workflow" class="md-nav__link">
    Gather Workflow
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#taxonomic-classification-with-kaiju" title="Taxonomic Classification with Kaiju" class="md-nav__link">
    Taxonomic Classification with Kaiju
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kaiju-report-workflow" title="Kaiju Report Workflow" class="md-nav__link">
    Kaiju Report Workflow
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaiju-filtered-species-report-workflow" title="Kaiju Filtered Species Report Workflow" class="md-nav__link">
    Kaiju Filtered Species Report Workflow
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaiju-filtered-species-by-class-report-workflow" title="Kaiju Filtered Species by Class Report Workflow" class="md-nav__link">
    Kaiju Filtered Species by Class Report Workflow
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="quick-start">Quick Start<a class="headerlink" href="#quick-start" title="Permanent link">&para;</a></h1>
<p>Let's run through the entire process start to finish.</p>
<h2 id="useful-snakemake-flags">Useful Snakemake Flags<a class="headerlink" href="#useful-snakemake-flags" title="Permanent link">&para;</a></h2>
<p>Two useful snakemake flags that you can add are:</p>
<ul>
<li><code>--dryrun</code> or <code>-n</code>: do a dry run of the workflow but do not actually run any commands</li>
<li><code>--printshellcmds</code> or <code>-p</code>: print the shell commands that are being executed
    (or would be executed if combined with <code>--dryrun</code>)</li>
</ul>
<h2 id="read-filtering">Read Filtering<a class="headerlink" href="#read-filtering" title="Permanent link">&para;</a></h2>
<p>We will run two variations of the read filtering workflow, and perform a quality
assessment of our reads both before and after quality trimming.</p>
<p>Before you begin, make sure you have Singularity installed as in
the <a href="../installing/">Installing</a> documentation.</p>
<p>Start by cloning a copy of the repository:</p>
<div class="codehilite"><pre><span></span>git clone https://github.com/dahak-metagenomics/dahak
</pre></div>


<p>then move into the <code>workflows/</code> directory of the Dahak repository:</p>
<div class="codehilite"><pre><span></span><span class="nb">cd</span> dahak/workflows/
</pre></div>


<p>Now create a JSON file that defines a Snakemake configuration dictionary.
This file should:</p>
<ul>
<li>Provide URLs at which each read filtering file can be accessed</li>
<li>Provide a set of quality trimming values to use (2 and 30)</li>
<li>Specify which read files should be used for the workflow</li>
<li>Specify a container image from the biocontainers project to use with
  Singluarity</li>
<li>Set all read filtering parameters</li>
</ul>
<p>(See the <a href="../readfilt_snakemake/">Read Filtering Snakemake</a> page for details on
these options.)</p>
<p>Copy and paste the following:</p>
<div class="codehilite"><pre><span></span>cat &gt; readfilt.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
    },

    &quot;workflows&quot; : {
        &quot;read_filtering_pretrim_workflow&quot; : {
            &quot;sample&quot;    : [&quot;SRR606249_subset10&quot;],
        },
        &quot;read_filtering_posttrim_workflow&quot; : {
            &quot;sample&quot;    : [&quot;SRR606249_subset10&quot;],
        },
    },
}
EOF
</pre></div>


<p>This creates a workflow configuration file <code>readfilt.json</code> that will
download the example data files and configure one or more workflows.</p>
<p>We will run two workflows: one pre-trimming
quality assessment, and one post-trimming quality assessment, so we call
Snakemake and pass it two build targets: <code>read_filtering_pretrim_workflow</code>
and <code>read_filtering_posttrim_workflow</code>.</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>readfilt.json <span class="se">\</span>
          read_filtering_pretrim_workflow read_filtering_posttrim_workflow
</pre></div>


<p>This command outputs FastQC reports for the untrimmed reads as well as
the reads trimmed at both quality cutoffs, and also outputs the trimmed
PE reads and the orphaned reads. All files are placed in the <code>data/</code>
subdirectory.</p>
<p><br />
<br /></p>
<h2 id="assembly">Assembly<a class="headerlink" href="#assembly" title="Permanent link">&para;</a></h2>
<p>We will run two assembler workflows using the Megahit assembler workflow
implemented in Dahak.</p>
<p>(See the <a href="../assembly_snakemake/">Assembly Snakemake</a> page for details on
these options.)</p>
<p>Create a JSON file that defines a Snakemake configuration dictionary:</p>
<div class="codehilite"><pre><span></span><span class="err">cat</span> <span class="err">&gt;</span> <span class="err">assembly.json</span> <span class="err">&lt;&lt;EOF</span>
<span class="p">{</span>
    <span class="nt">&quot;files&quot;</span> <span class="p">:</span> <span class="p">{</span>
        <span class="nt">&quot;SRR606249_subset10_1_reads.fq.gz&quot;</span> <span class="p">:</span>  <span class="s2">&quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;</span><span class="p">,</span>
        <span class="nt">&quot;SRR606249_subset10_2_reads.fq.gz&quot;</span> <span class="p">:</span>  <span class="s2">&quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;</span><span class="p">,</span>
    <span class="p">},</span>

    <span class="nt">&quot;workflows&quot;</span> <span class="p">:</span> <span class="p">{</span>
        <span class="nt">&quot;assembly_workflow_megahit&quot;</span> <span class="p">:</span> <span class="p">{</span>
            <span class="nt">&quot;sample&quot;</span>    <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SRR606249_subset10&quot;</span><span class="p">],</span>
            <span class="nt">&quot;qual&quot;</span>      <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;2&quot;</span><span class="p">,</span><span class="s2">&quot;30&quot;</span><span class="p">],</span>
        <span class="p">}</span>
    <span class="p">},</span>
<span class="p">}</span>
<span class="err">EOF</span>
</pre></div>


<p>To run the assembly workflow with
both assemblers, we call Snakemake with the <code>assembly_workflow_all</code> target.</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>assembly.json <span class="se">\</span>
          assembly_workflow_megahit
</pre></div>


<p><br />
<br /></p>
<h2 id="comparison">Comparison<a class="headerlink" href="#comparison" title="Permanent link">&para;</a></h2>
<p>In this section we will run a comparison workflow to compute sourmash
signatures for both filtered reads and assemblies, and compare the
computed signatures to a reference database.</p>
<p>Create a config file:</p>
<p>(See the <a href="../comparison_snakemake/">Comparison Snakemake</a> page for details on
these options.)</p>
<p>Copy and paste the following:</p>
<div class="codehilite"><pre><span></span>cat &gt; comparison.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;
    },

    &quot;workflows&quot; : {
        &quot;comparison_workflow_reads&quot; : {
            &quot;kvalue&quot;    : [&quot;21&quot;,&quot;31&quot;,&quot;51&quot;],
        }
    },
}
EOF
</pre></div>


<p>Now, run the <code>comparison_workflow_reads</code> workflow:</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>comparison.json <span class="se">\</span>
          comparison_workflow_reads
</pre></div>


<p><br />
<br /></p>
<h2 id="taxonomic-classification">Taxonomic Classification<a class="headerlink" href="#taxonomic-classification" title="Permanent link">&para;</a></h2>
<h3 id="taxonomic-classification-with-sourmash">Taxonomic Classification with Sourmash<a class="headerlink" href="#taxonomic-classification-with-sourmash" title="Permanent link">&para;</a></h3>
<p>There are a number of taxonomic classification workflows
implemented in Dahak. In this section we cover the use
of the sourmash tool for taxonomic classification.</p>
<p>Before you begin, make sure you have everything listed on the
<a href="../installing/">Installing</a> page available on your command line.</p>
<p>There are two taxonomic classification build rules that use sourmash:
<code>taxonomic_classification_signatures_workflow</code> and
<code>taxonomic_classification_gather_workflow</code>.</p>
<h4 id="signatures-workflow">Signatures Workflow<a class="headerlink" href="#signatures-workflow" title="Permanent link">&para;</a></h4>
<p>The signatures workflow uses sourmash to compute k-mer signatures from read
files. This is essentially the same as the compute signatures step in the
comparison workflow.</p>
<p>(See the <a href="../taxclass_snakemake/">Taxonomic Classification Snakemake</a> page for
details on this workflow.)</p>
<div class="codehilite"><pre><span></span>cat &gt; compute.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
    },

    &quot;workflows&quot; : {
        &quot;taxonomic_classification_signatures_workflow&quot; : {
            &quot;sample&quot;  : [&quot;SRR606249_subset10&quot;],
            &quot;qual&quot; : [&quot;2&quot;,&quot;30&quot;]
        }
    }
}
EOF
</pre></div>


<div class="codehilite"><pre><span></span>export SINGULARITY_BINDPATH=&quot;data:/data&quot;

snakemake --use-singularity \
        --configfile=compute.json \
        taxonomic_classification_signatures_workflow
</pre></div>


<h4 id="gather-workflow">Gather Workflow<a class="headerlink" href="#gather-workflow" title="Permanent link">&para;</a></h4>
<p>The gather workflow uses sourmash to (gather?) signatures 
computed from read files and compare them to signatures stored
in a genome database.</p>
<p>Create a JSON file for the taxonomic classification gather workflow 
that defines a Snakemake configuration dictionary:</p>
<div class="codehilite"><pre><span></span>cat &gt; gather.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
    },

    &quot;workflows&quot; : {
        &quot;taxonomic_classification_gather_workflow&quot; : {
            &quot;sample&quot;  : [&quot;SRR606249_subset10&quot;],
            &quot;qual&quot; : [&quot;2&quot;,&quot;30&quot;]
        }
    }
}
EOF
</pre></div>


<p>To run the gather workflow, we call Snakemake with the
<code>taxonomic_classification_gather_workflow</code> target.</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>taxkaiju.json <span class="se">\</span>
          taxonomic_classification_gather_workflow
</pre></div>


<h3 id="taxonomic-classification-with-kaiju">Taxonomic Classification with Kaiju<a class="headerlink" href="#taxonomic-classification-with-kaiju" title="Permanent link">&para;</a></h3>
<p>There are several taxonomic classification workflows in Dahak that 
use the Kaiju tool as well. This section covers those workflows.</p>
<p>There are three taxonomic classification build rules that use kaiju:</p>
<ul>
<li><code>taxonomic_classification_kaijureport_workflow</code></li>
<li><code>taxonomic_classification_kaijureport_filtered_workflow</code></li>
<li><code>taxonomic_classification_kaijureport_filteredclass_workflow</code></li>
</ul>
<h4 id="kaiju-report-workflow">Kaiju Report Workflow<a class="headerlink" href="#kaiju-report-workflow" title="Permanent link">&para;</a></h4>
<p>Create a JSON file that defines a Snakemake configuration dictionary:</p>
<div class="codehilite"><pre><span></span>cat &gt; taxkaiju.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
    },

    &quot;workflows&quot; : {
        &quot;taxonomic_classification_kaijureport_workflow&quot; : {
            &quot;sample&quot;  : [&quot;SRR606249_subset10&quot;],
            &quot;qual&quot; : [&quot;2&quot;,&quot;30&quot;]
        }
    }
}
EOF
</pre></div>


<p>To run the taxonomic classification workflow
to generate a kaiju report, we call Snakemake with the
<code>taxonomic_classification</code> target.</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>taxkaiju.json <span class="se">\</span>
          taxonomic_classification_kaijureport_workflow
</pre></div>


<h4 id="kaiju-filtered-species-report-workflow">Kaiju Filtered Species Report Workflow<a class="headerlink" href="#kaiju-filtered-species-report-workflow" title="Permanent link">&para;</a></h4>
<p>The filtered kaiju workflow filters for species whose reads
compose less than N% of the total reads, where N is a parameter
set by the user. </p>
<p>Copy and paste the following:</p>
<div class="codehilite"><pre><span></span>cat &gt; taxkaiju_filtered.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
    },

    &quot;taxonomic_classification&quot; : {
        &quot;filter_taxa&quot; : {
            &quot;pct_threshold&quot; : 1
        }
    },

    &quot;workflows&quot; : {
        &quot;taxonomic_classification_kaijureport_filtered_workflow&quot; : {
            &quot;sample&quot;  : [&quot;SRR606249_subset10&quot;],
            &quot;qual&quot; : [&quot;2&quot;,&quot;30&quot;]
        }
    }
}
EOF
</pre></div>


<p>To run the taxonomic classification filtered report workflow,
we call Snakemake with the
<code>taxonomic_classification_kaijureport_filtered_workflow</code> target.</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>taxkaiju_filtered.json <span class="se">\</span>
          taxonomic_classification_kaijureport_filtered_workflow
</pre></div>


<h4 id="kaiju-filtered-species-by-class-report-workflow">Kaiju Filtered Species by Class Report Workflow<a class="headerlink" href="#kaiju-filtered-species-by-class-report-workflow" title="Permanent link">&para;</a></h4>
<p>The last workflow implements filtering but also implements reporting
the taxa level reported by kaiju. This iuses the "genus" taxonomic rank
level by default.</p>
<p>Copy and paste the following:</p>
<div class="codehilite"><pre><span></span>cat &gt; taxkaiju_filteredclass.json &lt;&lt;EOF
{
    &quot;files&quot; : {
        &quot;SRR606249_subset10_1_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
        &quot;SRR606249_subset10_2_reads.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
    },

    &quot;taxonomic_classification&quot; : {
        &quot;kaiju_report&quot; : {
            &quot;taxonomic_rank&quot; : &quot;genus&quot;
        }
    },

    &quot;workflows&quot; : {

        &quot;taxonomic_classification_kaijureport_filteredclass_workflow&quot; : {
            &quot;sample&quot;  : [&quot;SRR606249_subset10&quot;],
            &quot;qual&quot; : [&quot;2&quot;,&quot;30&quot;]
        },
    }
}
</pre></div>


<p>To run the taxonomic classification workflow
to generate this kaiju report, we call Snakemake with the
<code>taxonomic_classification_kaijureport_filteredclass_workflow</code> target.</p>
<div class="codehilite"><pre><span></span><span class="nb">export</span> <span class="nv">SINGULARITY_BINDPATH</span><span class="o">=</span><span class="s2">&quot;data:/data&quot;</span>

snakemake --use-singularity <span class="se">\</span>
          --configfile<span class="o">=</span>taxkaiju_filtered.json <span class="se">\</span>
          taxonomic_classification_kaijureport_filteredclass_workflow
</pre></div>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../config/" title="Workflow Configuration" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Workflow Configuration
              </span>
            </div>
          </a>
        
        
          <a href="../troubleshooting/" title="Troubleshooting" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Troubleshooting
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2017-2018, The Regents of the University of California.<br /> Written and maintained by the <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a> at UC Davis.<br /> Released under the <a href="https://opensource.org/licenses/BSD-3-Clause">BSD 3-Clause License</a>.<br /><br />
          </div>
        
        powered by
        <a href="http://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.0cf9b500.js"></script>
      
      <script>app.initialize({version:"0.17.5",url:{base:".."}})</script>
      
        <script src="../search/require.js"></script>
      
        <script src="../search/search.js"></script>
      
    
    
      
    
  </body>
</html>